src/protocol/vm/state_builder.rs

```diff
@@ -0,0 +1,461 @@
+        Self {
+            id,
+            tokens,
+            block,
+            balances: None,
+            balance_owner: None,
+            capabilities: None,
+            involved_contracts: None,
+            stateless_contracts: None,
+            token_storage_slots: None,
+            manual_updates: None,
+            trace: None,
+            engine: None,
+            adapter_contract: None,
+            adapter_contract_path: None,
+        }
+    }
+
+    pub fn balances(mut self, balances: HashMap<H160, U256>) -> Self {
+        self.balances = Some(balances);
+        self
+    }
+
+    pub fn balance_owner(mut self, balance_owner: H160) -> Self {
+        self.balance_owner = Some(balance_owner);
+        self
+    }
+
+    pub fn capabilities(mut self, capabilities: HashSet<Capability>) -> Self {
+        self.capabilities = Some(capabilities);
+        self
+    }
+
+    pub fn involved_contracts(mut self, involved_contracts: HashSet<H160>) -> Self {
+        self.involved_contracts = Some(involved_contracts);
+        self
+    }
+
+    pub fn stateless_contracts(
+        mut self,
+        stateless_contracts: HashMap<String, Option<Vec<u8>>>,
+    ) -> Self {
+        self.stateless_contracts = Some(stateless_contracts);
+        self
+    }
+
+    pub fn token_storage_slots(
+        mut self,
+        token_storage_slots: HashMap<H160, (ERC20Slots, ContractCompiler)>,
+    ) -> Self {
+        self.token_storage_slots = Some(token_storage_slots);
+        self
+    }
+
+    pub fn manual_updates(mut self, manual_updates: bool) -> Self {
+        self.manual_updates = Some(manual_updates);
+        self
+    }
+
+    pub fn trace(mut self, trace: bool) -> Self {
+        self.trace = Some(trace);
+        self
+    }
+
+    pub fn engine(mut self, engine: SimulationEngine<PreCachedDB>) -> Self {
+        self.engine = Some(engine);
+        self
+    }
+
+    pub fn adapter_contract(
+        mut self,
+        adapter_contract: TychoSimulationContract<PreCachedDB>,
+    ) -> Self {
+        self.adapter_contract = Some(adapter_contract);
+        self
+    }
+
+    pub fn adapter_contract_path(mut self, adapter_contract_path: String) -> Self {
+        self.adapter_contract_path = Some(adapter_contract_path);
+        self
+    }
+
+    /// Build the final VMPoolState object
+    pub async fn build(mut self) -> Result<VMPoolState<PreCachedDB>, SimulationError> {
+        let engine = if let Some(engine) = &self.engine {
+            engine.clone()
+        } else {
+            self.get_default_engine().await?
+        };
+
+        if self.adapter_contract.is_none() {
+            self.adapter_contract = Some(TychoSimulationContract::new_swap_adapter(
+                *ADAPTER_ADDRESS,
+                self.adapter_contract_path
+                    .as_ref()
+                    .ok_or_else(|| {
+                        SimulationError::NotInitialized("Adapter contract".to_string())
```

Comment:
Not really, if we're here it means that no adapter contract nor adapter contract path were given, this means we're unable to initialize the adapter contract because we miss some information.

Also ideally we would like to enforce that at least one of them is given at compile time, but it adds some generics and therefore some complexity. So I decided to not do it for now but that's something I'm keeping in mind.